crates/sources/src/subgraph.rs

```diff
@@ -56,6 +68,45 @@ impl SubgraphClient {
             .await?
             .into_result()
     }
+
+    /// Performs the specified GraphQL query on the current subgraph.
+    /// This function should be called for queries that return very long(paginated) result.
+    pub async fn paginated_query<T>(&self, block_number: u64, query: &str) -> Result<Vec<T>>
+    where
+        T: ContainsId + DeserializeOwned,
+    {
+        let mut result = Vec::new();
+        let mut last_id = String::default();
+
+        // We do paging by last ID instead of using `skip`. This is the
+        // suggested approach to paging best performance:
+        // <https://thegraph.com/docs/en/developer/graphql-api/#pagination>
+        loop {
+            let page = self
+                .query::<Data<T>>(
+                    query,
+                    Some(json_map! {
+                        "block" => block_number,
+                        "pageSize" => QUERY_PAGE_SIZE,
+                        "lastId" => json!(last_id),
+                    }),
+                )
+                .await?
+                .inner;
+            let no_more_pages = page.len() != QUERY_PAGE_SIZE;
```

Comment:
nit: Unless thegraph messes up this shouldn't be a problem but maybe we should be defensive and use `page.len() < QUERY_PAGE_SIZE` here?

Reply:
Hm, not sure about this one, but it seems to me that the `!=` is more defensive than `<`. Anyway since nit, I would keep is as is since this particular line of code is taken from balancer and already in use for some time without any issues.

Reply:
Yeah, you're right.`<` would be actually more greedy not defensive.
I'm fine with both. 👌